src/function.rs

```diff
@@ -256,3 +256,12 @@ where
             .finish()
     }
 }
+
+// Consider replacing with `Try` when it stabilizes.
+pub trait HasOutput {
+    type Output;
+}
+
+impl<T, E> HasOutput for std::result::Result<T, E> {
+    type Output = T;
+}
```

Comment:
It is too bad that Specialization is a nightly-only feature. It would allow us to implement `HasOutput` for `T` and specialize it for `Result`. It would allow the macro to support both non-result and result return types when adding a new `into_result(self) -> salsa::Result<Self::Output>` method (which is a no-op for `salsa::Result` and returns `Ok(self)` for all other types.